kbkdf/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+}
+
+impl<Prf, K, R> Default for Counter<Prf, K, R> {
+    fn default() -> Self {
+        Self {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<Prf, K, R> Kbkdf<Prf, K, R> for Counter<Prf, K, R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+    R: sealed::R,
+{
+}
+
+pub struct Feedback<'a, Prf, K, R = U32>
+where
+    Prf: Mac,
+{
+    iv: Option<&'a Array<u8, Prf::OutputSize>>,
+    _marker: PhantomData<(Prf, K, R)>,
+}
+
+impl<'a, Prf, K, R> Feedback<'a, Prf, K, R>
+where
+    Prf: Mac,
+{
+    pub fn new(iv: Option<&'a Array<u8, Prf::OutputSize>>) -> Self {
+        Self {
+            iv,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<'a, Prf, K, R> Kbkdf<Prf, K, R> for Feedback<'a, Prf, K, R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+    R: sealed::R,
+{
+    fn input_iv(&self, ki: &mut Option<Array<u8, Prf::OutputSize>>) {
+        if let Some(iv) = self.iv {
+            *ki = Some(iv.clone())
+        }
+    }
+
+    const FEEDBACK_KI: bool = true;
+}
+
+pub struct DoublePipeline<Prf, K, R = U32>
+where
+    Prf: Mac,
+{
+    _marker: PhantomData<(Prf, K, R)>,
+}
+
+impl<Prf, K, R> Default for DoublePipeline<Prf, K, R>
+where
+    Prf: Mac,
+{
+    fn default() -> Self {
+        Self {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<Prf, K, R> Kbkdf<Prf, K, R> for DoublePipeline<Prf, K, R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+    R: sealed::R,
+{
+    const DOUBLE_PIPELINE: bool = true;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Array, Counter, DoublePipeline, Feedback, Kbkdf};
+    use core::convert::TryFrom;
+    use digest::{consts::*, crypto_common::KeySizeUser};
+    use hex_literal::hex;
+
+    #[derive(Debug)]
+    struct KnownValue {
+        key: &'static [u8],
+        iv: Option<&'static [u8]>,
+        use_l: bool,
+        label: &'static [u8],
+        context: &'static [u8],
+        use_separator: bool,
+        expected: &'static [u8],
+    }
+
+    static KNOWN_VALUES_COUNTER_HMAC_SHA256: &[KnownValue] = &[
+        KnownValue {
+            iv: None,
+            use_l: false,
+            use_separator: false,
+            label: &[],
+            context: &[],
+            key: &hex!(
+                "
+                241C3FBAABEDE87601B1C778B24F9A32 742A14FE34DA61D77E8352EF9D6C7FC8
+                E335E32344E21D7DC0CD627D7E2FF973 992611F372C5D3DD91C100F2C6DB2CAF
+            "
```

Comment:
You can write these constants like this:
```rust
&hex!(
    "241C3FBAABEDE87601B1C778B24F9A32"
    "742A14FE34DA61D77E8352EF9D6C7FC8"
    "E335E32344E21D7DC0CD627D7E2FF973"
    "992611F372C5D3DD91C100F2C6DB2CAF"
)
```

Reply:
https://github.com/TheBestTvarynka/KDFs/pull/4

Reply:
done